### PR TITLE
[quant] Add dequantize.tensors

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35130 [quant][graphmode] Make interpolate/upsample work again
* #35142 [quant][graphmode] SwapDeQuant support prim::If
* #35141 [quant][graphmode][refactor] getGeneralOpTensorInputIndexes -> getGeneralOpTensorInputs
* #35135 [quant][graphmode][refactor] swapDeQuant takes block as arugment
* #35077 [quant][graphmode] Fold quantized prepacking ops
* #35073 [quant] Make conv2d_prepack and linear_prepack pure
* #34855 [quant][graphmode] quantization support for prim::CallFunction
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34854 [quant][graphmode] quantization support for view, transpose, contiguos
* #34806 [quant][graphmode] quantization support for aten::chunk
* #34807 [quant][graphmode] quantization support for prim::ListUnpack
* #34804 [quant][graphmode] Replicate quantize node for prim::If
* #34572 [quant][graphmode] quantization support for aten::add
* #34518 [quant][graphmode] quantization work for prim::If
* #34571 [quant][graphmode][fix] use observed_values_ to check values are observed
* #34414 [quant][graphmode][refactor] insertObservers for Block
* #34411 [quant][graphmode][fix] Insert dequantize before use node
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node
* **#34348 [quant] Add dequantize.tensors**

Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

### CODE REVIEW

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 1]
```diff
@@ -32,6 +32,14 @@ Tensor dequantize_quant(const Tensor& self) {
   return get_qtensorimpl(self)->quantizer()->dequantize(self);
 }
 
+std::vector<Tensor> dequantize_tensors_quant(TensorList tensors) {
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/QTensor.cpp]: Why is this needed? Can we handle lists in a more primitive manner? What if we need to add a dict of tensors, would we add one more function for this purpose?

In general, how are lists of tensors handled? How do we observe a list of tensors and quantize them?

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/QTensor.cpp]: explained offline

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -3601,12 +3601,17 @@
...
     CPU: quantize_per_channel_cpu
 
-- func: dequantize(Tensor self) -> Tensor
+- func: dequantize.self(Tensor self) -> Tensor
   use_c10_dispatcher: full
   variants: function, method
   dispatch:
     QuantizedCPU: dequantize_quant
 
+- func: dequantize.tensors(Tensor[] tensors) -> Tensor[]
```
Comment (supriyar) [inline aten/src/ATen/native/native_functions.yaml]: Why is ListConstruct special? What about other structures like Tuple or Dict of tensors?

Comment (jerryzh168) [reply aten/src/ATen/native/native_functions.yaml]: we'll need to handle those as well if they come up in real use cases, but by far I didn't see that. Reason we have lists are mostly because we have quantized ops like cat that takes list as input. if we have quantized ops that takes Dict or Tuple as inputs we'll probably need to add support for these other types as well

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 3]
```diff
@@ -32,6 +32,14 @@ Tensor dequantize_quant(const Tensor& self) {
...
+  std::vector<Tensor> dequantized_tensors;
+  for (auto i = 0; i < tensors.size(); ++i) {
```
Comment (z-a-f) [inline aten/src/ATen/native/quantized/QTensor.cpp]: can we use parallel for here?

**Review by z-a-f — APPROVED** [review]
LGTM

File: aten/src/ATen/native/quantized/QTensor.cpp (hunk 3 above)
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/QTensor.cpp]: I would delay this optimization until there is a need, most likely this function will never be run since it will be fused into the quantized::cat pattern.

File: aten/src/ATen/native/quantized/QTensor.cpp (hunk 1 above)
Comment (smessmer) [inline aten/src/ATen/native/quantized/QTensor.cpp]: Don't return a std::vector, return a TensorList instead (TensorList == torch::List<Tensor>). std::vector requires a copy behind the scenes.

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -1513,6 +1513,20 @@ def merge_dicts(*dicts):
...
+Given a quantized Tensor, dequantize it and return the dequantized float Tensor.
+""".format(**reduceops_common_args))
+
+add_docstr(torch.dequantize,
+           r"""
+dequantize(tensors) -> List of Tensors
+
+Given a list of quantized Tensors, dequantize them and return a list of dequantized Tensors
+""".format(**reduceops_common_args))
+
```
Comment (jerryzh168) [inline torch/_torch_docs.py]: This seems impossible, what should I do here? @smessmer ?

File: aten/src/ATen/native/quantized/QTensor.cpp (hunk 1 above)
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/QTensor.cpp]: I think that is not the type generated by native_functions.yaml, I looked up another function that returns Tensor[] and it says:
```
795:std::vector<Tensor> meshgrid(TensorList tensors); // {"schema": "aten::meshgrid(Tensor[] tensors) -> Tensor[]", "compound": "true"}
```
